Framework/API/inc/MantidAPI/WorkspacePropertyWithIndex.h

```diff
@@ -0,0 +1,122 @@
+#ifndef MANTID_API_WORKSPACEPROPERTYWITHINDEX_H_
+#define MANTID_API_WORKSPACEPROPERTYWITHINDEX_H_
+
+#include "MantidAPI/IWorkspacePropertyWithIndex.h"
+#include "MantidAPI/WorkspaceProperty.h"
+#include <MantidAPI/IndexTypeProperty.h>
+#include <MantidIndexing/GlobalSpectrumIndex.h>
+#include <MantidIndexing/SpectrumIndexSet.h>
+#include <MantidIndexing/SpectrumNumber.h>
+#include <MantidKernel/ArrayProperty.h>
+#include <algorithm>
+#include <boost/shared_ptr.hpp>
+#include <memory>
+#include <tuple>
+#include <vector>
+
+namespace Mantid {
+namespace API {
+class MatrixWorkspace;
+
+/** WorkspacePropertyWithIndex : Workspace property which allows users to
+  specify an input index type. The property then simultaneously returns the
+  workspace and the indices.
+  @author Lamar Moore
+  @Date 05-05-2017
+  Copyright &copy; 2017 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+  National Laboratory & European Spallation Source
+
+  This file is part of Mantid.
+
+  Mantid is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 3 of the License, or
+  (at your option) any later version.
+
+  Mantid is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+template <typename TYPE = MatrixWorkspace>
+class WorkspacePropertyWithIndex : public WorkspaceProperty<TYPE>,
```

Comment:
@SimonHeybrock it would. We essentially need to re-implement a `PropertyWithValue<T>` which is not worth it.